src/traversal/bfs_visit.rs

```diff
@@ -0,0 +1,67 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use pyo3::prelude::*;
+
+use petgraph::stable_graph::NodeIndex;
+use petgraph::visit::Control;
+
+use crate::PruneSearch;
+use retworkx_core::traversal::BfsEvent;
+
+#[derive(FromPyObject)]
+pub struct PyBfsVisitor {
+    discover_vertex: PyObject,
+    finish_vertex: PyObject,
+    tree_edge: PyObject,
+    non_tree_edge: PyObject,
+    gray_target_edge: PyObject,
+    black_target_edge: PyObject,
+}
+
+pub fn bfs_handler(
+    py: Python,
+    vis: &PyBfsVisitor,
+    event: BfsEvent<NodeIndex, &PyObject>,
+) -> PyResult<Control<()>> {
+    let res = match event {
+        BfsEvent::Discover(u) => vis.discover_vertex.call1(py, (u.index(),)),
+        BfsEvent::TreeEdge(u, v, weight) => {
+            let edge = (u.index(), v.index(), weight);
+            vis.tree_edge.call1(py, (edge,))
+        }
+        BfsEvent::NonTreeEdge(u, v, weight) => {
+            let edge = (u.index(), v.index(), weight);
+            vis.non_tree_edge.call1(py, (edge,))
+        }
+        BfsEvent::GrayTargetEdge(u, v, weight) => {
+            let edge = (u.index(), v.index(), weight);
+            vis.gray_target_edge.call1(py, (edge,))
+        }
+        BfsEvent::BlackTargetEdge(u, v, weight) => {
+            let edge = (u.index(), v.index(), weight);
+            vis.black_target_edge.call1(py, (edge,))
+        }
+        BfsEvent::Finish(u) => vis.finish_vertex.call1(py, (u.index(),)),
+    };
+
+    match res {
+        Err(e) => {
+            if e.is_instance::<PruneSearch>(py) {
+                Ok(Control::Prune)
+            } else {
+                Err(e)
+            }
```

Comment:
So `StopSearch` isn't special here? I assumed  you added a custom class so it would be treated as a special exit condition  but this just raises. If this is the case do we need a special stop exception?

Reply:
Yeah, it's not really special but just defined for user convenience. It's treated as any other exception that may occur in python side, i.e exit early.

Reply:
I'm wondering if we want to make it special, so if a user callback raises `StopSearch` it causes the function to return without an error kind of like how  `for i in iter` doesn't raise `StopIteration` when iter is exhausted. That was at least my expectation when I saw the 2 exception classes.

Reply:
Sure, that's a good point. My main thought was if we do not raise, a user might not know if the search terminated normally or exited early and a further action might depend on that (as in [#7211](https://github.com/Qiskit/qiskit-terra/pull/7211)) . But I guess users can define their own custom exception if they wish this kind of behavior.

Reply:
Done in 871d9eb.